lib/_http_agent.js

```diff
@@ -202,12 +202,23 @@ Agent.prototype.addRequest = function addRequest(req, options, port/* legacy */,
     this.sockets[name] = [];
   }
 
-  const freeLen = this.freeSockets[name] ? this.freeSockets[name].length : 0;
+  const freeSockets = this.freeSockets[name];
+  let socket;
+  while (!socket && freeSockets && freeSockets.length) {
+    socket = freeSockets.shift();
+    // TODO(ronag): writable, _writabeState.errored?
+    if (socket.destroyed) {
+      socket = null;
+    }
+    // don't leak
+    if (!freeSockets.length)
+      delete this.freeSockets[name];
+  }
```

Comment:
Nit: a bit less array modifications and state
```js
const freeSockets = this.freeSockets[name];
if (freeSockets) {
  let skip = 0;
  while (skip < freeSockets.length && freeSockets[skip].destroyed)
    skip++;
  if (skip > 0) {
    if (skip < freeSockets.length) {
      freeSockets.splice(0, skip);
    } else {
      freeSockets.length = 0;
      delete this.freeSockets[name];
    }
  }
}

const freeLen = freeSockets ? freeSockets.length : 0;
const sockLen = freeLen + this.sockets[name].length;

if (freeLen > 0)
```

OR if we don't expect there to be many `destroyed` sockets (which is probably the case) I think this will be fine:
```js
const freeSockets = this.freeSockets[name];
if (freeSockets) {
  while (freeSockets.length > 0) {
    if (freeSockets[0].destroyed) freeSockets.shift();
  }
  if (freeSockets.length == 0) {
    delete this.freeSockets[name];
  }
}
```

Reply:
I don't think that quite works. I would prefer to keep it mostly as is.

Reply:
I did a minor simplification that you might find more agreeable.